tests/Meziantou.Analyzer.Test/Helpers/ProjectBuilder.Validation.cs

```diff
@@ -453,6 +475,34 @@ private static async Task<string> GetStringFromDocument(Document document)
             root = Formatter.Format(root, Formatter.Annotation, simplifiedDoc.Project.Solution.Workspace);
             return root.GetText().ToString();
         }
+
+        private sealed class CustomDiagnosticProvider : FixAllContext.DiagnosticProvider
+        {
+            private readonly Diagnostic[] _diagnostics;
+
+            public CustomDiagnosticProvider(Diagnostic[] diagnostics)
+            {
+                _diagnostics = diagnostics;
+            }
+
+            public override Task<IEnumerable<Diagnostic>> GetAllDiagnosticsAsync(Project project, CancellationToken cancellationToken)
+            {
+                // TODO I'm not sure of this one
```

Comment:
Now that you have implemented a FixAllProvider, do you know if this implementation is ok?

Reply:
I'll have a look...

Reply:
I made a few changes. Let me know what you think.

Reply:
I'm not sure it is valid to check the `FilePath`. As indicated [in the comment](https://github.com/dotnet/roslyn/blob/2869fb77525e6128243e7607760c2dbe75864721/src/Compilers/Core/Portable/Syntax/SyntaxTree.cs#L38), the path can be empty if the source is generated and the value may not be unique. It may be safer to compare the syntaxtree or syntaxroot.

Reply:
Despite the Roslyn comment, I felt that using `FilePath` here was not a bad idea, because:
1. We control what gets stored in the `Document` objects we create (and I made sure we stored a non-empty FilePath)
1. [StyleCop also uses FilePath](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/ce8ee2f198f6138b04b968ffb73e1dc6d6e66686/StyleCop.Analyzers/StyleCopTester/TesterDiagnosticProvider.cs#L51) in their tests to determine whether a `Diagnostic` corresponds to a given `Document`. Note, however, that they make a distinction between `Project`-specific Diagnostics and `Document`-specific ones.

Nevertheless, I can go back to using the SyntaxTree root.